### PR TITLE
Add StochasticDepth implementation

### LINKED ISSUE #4298: Implement a Stochastic Depth operator
## 🚀 Feature
Implement the Stochastic Depth as described in [Deep Networks with Stochastic Depth](https://arxiv.org/abs/1603.09382) to randomly drop residual branches in residual architectures.

The operator should support 2 modes, dropping the entire batch as described on the original paper and dropping individual rows from the batch as utilized on the [EfficientNet paper](https://arxiv.org/pdf/1905.11946.pdf). We should provide both a functional and a Layer form.

### PR DESCRIPTION
**Pull request by datumbox**
Fixes #4298

### CODE REVIEW

File: test/test_ops.py [hunk 1]
```diff
@@ -1000,5 +1006,34 @@ def gen_iou_check(box, expected, tolerance=1e-4):
             gen_iou_check(box_tensor, expected, tolerance=0.002 if dtype == torch.float16 else 1e-3)
 
 
+class TestStochasticDepth:
+    @pytest.mark.skipif(stats is None, reason="scipy.stats not available")
```
Comment (NicolasHug) [inline test/test_ops.py:1010]: As a slightly cleaner pattern we could use: `pexpect = pytest.importorskip("pexpect")` in the test body https://docs.pytest.org/en/latest/how-to/skipping.html#summary

But is there a reason we should not expect stats to be available here? I think we rely on scipy unconditionally in other tests

File: test/test_ops.py [hunk 2]
```diff
@@ -1000,5 +1006,34 @@ def gen_iou_check(box, expected, tolerance=1e-4):
...
+    @pytest.mark.parametrize('mode', ["batch", "row"])
+    @pytest.mark.parametrize('p', [0.2, 0.5, 0.8])
+    def test_stochastic_depth(self, mode, p):
+        random.seed(42)
```
Comment (NicolasHug) [inline test/test_ops.py:1014]: Shouldn't this be the torch RNG that we want to fix? I see calls to `torch.rand` in `stochastic_depth()`

File: torchvision/ops/stochastic_depth.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    Implements the Stochastic Depth from `"Deep Networks with Stochastic Depth"
+    <https://arxiv.org/abs/1603.09382>`_ used for randomly dropping residual
+    branches of residual architectures.
+
+    Args:
+        input (Tensor[N, ...]): The input tensor or arbitrary dimensions with the first one
+                    being its batch i.e. a batch with ``N`` rows.
+        mode (str): ``"batch"`` or ``"row"``.
+                    ``"batch"`` randomly zeroes the entire input, ``"row"`` zeroes
+                    randomly selected rows from the batch.
```
Comment (NicolasHug) [inline torchvision/ops/stochastic_depth.py:16]: Do I understand correctly that with "batch" the zero-ing is done element-wise, e.g. a row can have zero and non-zero elements, whereas with "row", some rows would be entirely zeroed?

Comment (NicolasHug) [reply torchvision/ops/stochastic_depth.py:16]: As a follow-up I'm wondering if there could be a more descriptive name instead of "row". I don't know in which context this would be used but for example if we know the input is always a batch of feature maps, this could be "featuremap" instead or something like that?

Again, depending on the context in which this is used, this comment may not make sense. 

Context: I'm unfamiliar with the use of the term "row" in a non-2D array, but perhaps it's a common generalization?

File: torchvision/ops/stochastic_depth.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    survival_rate = 1.0 - p
+    if mode == "batch":
+        keep = torch.rand(size=(1, ), dtype=input.dtype, device=input.device) < survival_rate
+    elif mode == "row":
+        keep = torch.rand(size=(input.size(0),), dtype=input.dtype, device=input.device) < survival_rate
+        keep = keep[(None, ) * (input.ndim - 1)].T
+    else:
+        raise ValueError("mode has to be either 'batch' or 'row', but got {}".format(mode))
+    return input / survival_rate * keep
```
Comment (NicolasHug) [inline torchvision/ops/stochastic_depth.py:36]: The `training` param doesn't seem to be used. Should we only devide by `survival_rate` when it's True?

**Review by NicolasHug — COMMENTED** [review]
Thanks @datumbox ! I made a brief review with some questions, but this looks good

File: torchvision/ops/stochastic_depth.py (hunk 4 above)
Comment (fmassa) [inline torchvision/ops/stochastic_depth.py:36]: nit: here are some potential micro-optimizations, as we don't need to divide the whole input by `survival_rate`, but we can do it only in the `keep`. This is similar to what is done in the [PyTorch implementation of dropout](https://github.com/pytorch/pytorch/blob/913c1f83f49f9e1e2a494186cc0069d780cee852/aten/src/ATen/native/Dropout.cpp#L62)

```suggestion
    if mode not in ["batch", "row"]:
        raise ValueError("mode has to be either 'batch' or 'row', but got {}".format(mode))
    size = [1] * input.ndim
    if mode == "row":
       size[0] = input.shape[0]
   noise = torch.empty(size, dtype=input.dtype, device=input.device)
   noise = noise.bernoulli_(survival_rate).div_(survival_rate)
   return input * noise
```

What I like about this implementation is that it makes it even clearer what the differences are between the different modes. But it's a minor difference in clarity, so I'm not set to stone on my proposal.

File: torchvision/ops/stochastic_depth.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
+import torch
+from torch import nn, Tensor
+
+
+def stochastic_depth(input: Tensor, mode: str, p: float, training: bool = True) -> Tensor:
```
Comment (fmassa) [inline torchvision/ops/stochastic_depth.py:5]: What do you think about making the signature be
```python
input: Tensor, p: float, mode: str, training: bool = True
```
instead? I think having `p` come first might be more natural?

File: torchvision/ops/stochastic_depth.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    else:
+        raise ValueError("mode has to be either 'batch' or 'row', but got {}".format(mode))
+    return input / survival_rate * keep
+
+
+class StochasticDepth(nn.Module):
+    """
+    See :func:`stochastic_depth`.
+    """
+    def __init__(self, mode: str, p: float) -> None:
```
Comment (fmassa) [inline torchvision/ops/stochastic_depth.py:43]: Same comment here regarding making `p` come first

**Review by fmassa — APPROVED** [review]
Looks great, thanks a lot!

I have some small comments, let me know what you think. None of those are merge-blocking, so I'm approving it.

Also, I think you need to add an entry in https://github.com/pytorch/vision/blob/master/docs/source/ops.rst for it to appear in the documentation

File: test/test_ops.py (hunk 2 above)
Comment (datumbox) [reply test/test_ops.py:1014]: That's what I thought as well but then I saw the refactored version of the transforms and followed the same approach:
https://github.com/pytorch/vision/blob/3fa2055b5a4807a8ffc577c53da180382cafd17c/test/test_transforms.py#L522-L524

Do you think this was a mistake introduced during the refactoring?

File: torchvision/ops/stochastic_depth.py (hunk 4 above)
Comment (datumbox) [reply torchvision/ops/stochastic_depth.py:36]: We use it, check few lines up. And no the formula is correct.

File: torchvision/ops/stochastic_depth.py (hunk 3 above)
Comment (datumbox) [reply torchvision/ops/stochastic_depth.py:16]: I also don't like the name. That's the best we came up with Brian and Francisco. If you find better names, happy to replace them. 

Currently batch drops the entire batch while row drops random records of the batch.

File: test/test_ops.py (hunk 1 above)
Comment (datumbox) [reply test/test_ops.py:1010]: I'm using the same pattern we use here:
https://github.com/pytorch/vision/blob/3fa2055b5a4807a8ffc577c53da180382cafd17c/test/test_transforms.py#L17-L20

https://github.com/pytorch/vision/blob/3fa2055b5a4807a8ffc577c53da180382cafd17c/test/test_transforms.py#L512

Perhaps we could create a ticket to refactor all of them?

File: test/test_ops.py (hunk 2 above)
Comment (NicolasHug) [reply test/test_ops.py:1014]: It looks like these lines were introduced in https://github.com/pytorch/vision/pull/3123

Perhaps I'm missing what this is supposed to do then: are we trying to fix the randomness of `StochasticDepth`, or the randomness of something else, and what is this "something else"?

As far as I understand the test procedure in itself (the call to scipy.stats) is deterministic?

File: torchvision/ops/stochastic_depth.py (hunk 4 above)
Comment (NicolasHug) [reply torchvision/ops/stochastic_depth.py:36]: From what I can tell we only use it here:

```py
    if not training or p == 0.0:
        return input
```

which makes me wonder whether this parameter is needed? If it's intended, I would suggest to update the docstring to:
> training: apply dropout if is ``True``. If ``False``, this transform is a no-op. Default: ``True``

File: test/test_ops.py (hunk 1 above)
Comment (NicolasHug) [reply test/test_ops.py:1010]: sure, but we might as well start using his already :)
We started using it in https://github.com/pytorch/vision/blob/master/test/test_onnx.py#L20 to skip an entire test file

File: test/test_ops.py (hunk 2 above)
Comment (datumbox) [reply test/test_ops.py:1014]: > It looks like these lines were introduced in #3123

This is before #3123. See [here](https://github.com/pytorch/vision/blame/4eab7a67f9f134c810eb7d1977e3ef9aa3fa1b26/test/test_transforms.py#L435-L445). Perhaps the intention was not to fix at all the rand on torch and only work on PIL/numpy randomness. I'll check if I can completely remove it.

> As far as I understand the test procedure in itself (the call to scipy.stats) is deterministic?
Yes, that's correct.

File: torchvision/ops/stochastic_depth.py (hunk 4 above)
Comment (datumbox) [reply torchvision/ops/stochastic_depth.py:36]: I'm following the same convention as in PyTorch's Dropout:
https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/dropout.py#L58

Comment (NicolasHug) [reply torchvision/ops/stochastic_depth.py:36]: Oh right, I forgot `self.training` gets set automatically. I'm wondering whether the current docstring is accurate though:

> training: apply dropout if is ``True``. Default: ``True``

Should `dropout` be something else?

File: test/test_ops.py (hunk 2 above)
Comment (datumbox) [reply test/test_ops.py]: It seems it passes without fixing the seed, so I removed it. If we see flakiness we can set the seed for both all randgens.

**Comment by datumbox** [comment]
Thanks both for the fast review. I believe I addressed all comments. I'll merge after the tests pass.
